File: Army(c++)/State/VampireState.cpp
#include "VampireState.hpp"
#include "../Basic/Unit.hpp"

VampireState::VampireState(Unit* thisUnit) : thisUnit(thisUnit) {}

void VampireState::changeUnit(Unit* enemy) {
    this->thisUnit->getHealth().ensureIsAlive();
    enemy->getHealth().ensureIsAlive();
    
    if ( enemy->getWarriorID() == VAMPIRE || enemy->getWarriorID() == WEREWOLF || enemy->getWarriorID() == WOLF || enemy->getWarriorID() == DEMON ) {
        return;
    }
    
    enemy->transitionTo(new VampireState(enemy));
    
    enemy->setWarriorID(WarriorID::VAMPIRE);
    enemy->getHealth().setDamage(this->thisUnit->getHealth().getDamage());
    enemy->getHealth().setHitPoints(this->thisUnit->getHealth().getHitPointsLimit());
    enemy->getHealth().setHitPointsLimit(this->thisUnit->getHealth().getHitPointsLimit());
}

void VampireState::transformTo() {
    std::cout << "No ability!!!"  << std::endl;
}

File: Army(java)/State/VampireState.java
package State;
import Basic.*;
import EnumID.*;

public class VampireState implements State {
    protected Unit thisUnit;

    public VampireState(Unit unit) {
        this.thisUnit = unit;
    }

    public void printState() {
        System.out.println("\tI'm Vampire!!!");
    }

    public void changeUnit(Unit enemy) throws UnitIsDeadException {
        thisUnit.getHealth().ensureIsAlive();
        enemy.getHealth().ensureIsAlive();
        
        if ( enemy.getWarriorID() == WarriorID.VAMPIRE || enemy.getWarriorID() == WarriorID.WEREWOLF || enemy.getWarriorID() == WarriorID.WOLF || enemy.getWarriorID() == WarriorID.DEMON ) {
            System.out.println("Error!!!");
            return;
        }
        
        enemy.setState(new VampireState(enemy));
        
        enemy.setWarriorID(WarriorID.VAMPIRE);
        enemy.getHealth().setDamage(thisUnit.getHealth().getDamage());
        enemy.getHealth().setHitPoints(thisUnit.getHealth().getHitPointsLimit());
        enemy.getHealth().setHitPointsLimit(thisUnit.getHealth().getHitPointsLimit());
    }

    public void transformTo() throws UnitIsDeadException {
        System.out.println("No ability!!!");
    }
}
